File: app/src/main/java/rssdemo/andras/hu/rssdemo/ui/subscriptions/editor/SubscriptionEditorViewModel.java
package rssdemo.andras.hu.rssdemo.ui.subscriptions.editor;


import rssdemo.andras.hu.rssdemo.R;
import rssdemo.andras.hu.rssdemo.data.Subscription;
import rssdemo.andras.hu.rssdemo.repository.SubscriptionRepository;

public class SubscriptionEditorViewModel {

    private SubscriptionRepository subscriptionRepository;
    private SubscriptionEditorView view;
    private boolean isEditMode;
    private String originalName;

    public SubscriptionEditorViewModel(SubscriptionRepository subscriptionRepository) {
        this.subscriptionRepository = subscriptionRepository;
    }

    public void setView(SubscriptionEditorView view, String originalName) {
        this.view = view;
        this.isEditMode = originalName != null;
        this.originalName = originalName == null ? "" : originalName;
    }

    public void onSaveClick() {
        Subscription subscription = view.getSubscrition();
        if (validateName(subscription.getName()) & validateUrl(subscription.getUrl())) {
            if (isEditMode) {
                subscriptionRepository.update(originalName, subscription);
            } else {
                subscriptionRepository.create(subscription);
            }
            view.closeEditor();
        }
    }

    private boolean validateName(String name) {
        if (name.length() == 0) {
            view.raiseErrorOnName(R.string.error_name_mandatory);
            return false;
        }
        if (!name.equals(originalName) && subscriptionRepository.isNameExisting(name)) {
            view.raiseErrorOnName(R.string.error_name_must_be_unique);
            return false;
        }
        return true;
    }

    private boolean validateUrl(String url) {
        if (url.length() == 0) {
            view.raiseErrorOnUrl(R.string.error_url_mandatory);
            return false;
        }
        return true;
    }

    public void onCancelClick() {
        view.closeEditor();
    }
}

File: app/src/main/java/rssdemo/andras/hu/rssdemo/ui/subscriptions/editor/SubscriptionEditorViewModel.kt
package rssdemo.andras.hu.rssdemo.ui.subscriptions.editor


import rssdemo.andras.hu.rssdemo.R
import rssdemo.andras.hu.rssdemo.repository.SubscriptionRepository

class SubscriptionEditorViewModel(private val subscriptionRepository: SubscriptionRepository) {
    private var isEditMode: Boolean = false
    private lateinit var view: SubscriptionEditorView
    private lateinit var originalName: String

    internal fun setView(view: SubscriptionEditorView, originalName: String) {
        this.view = view
        this.isEditMode = originalName.isNotEmpty()
        this.originalName = originalName
    }

    fun onSaveClick() {
        val subscription = view.getSubscription()
        if (validateName(subscription.name) and validateUrl(subscription.url)) {
            if (isEditMode) {
                subscriptionRepository.update(originalName, subscription)
            } else {
                subscriptionRepository.create(subscription)
            }
            view.closeEditor()
        }
    }

    private fun validateName(name: String): Boolean {
        if (name.isEmpty()) {
            view.raiseErrorOnName(R.string.error_name_mandatory)
            return false
        }
        if (name != originalName && subscriptionRepository.isNameExisting(name)) {
            view.raiseErrorOnName(R.string.error_name_must_be_unique)
            return false
        }
        return true
    }

    private fun validateUrl(url: String): Boolean {
        if (url.isEmpty()) {
            view.raiseErrorOnUrl(R.string.error_url_mandatory)
            return false
        }
        if (!url.startsWith("http://") && !url.startsWith("https://")) {
            view.raiseErrorOnUrl(R.string.error_not_url)
            return false
        }
        return true
    }

    fun onCancelClick() {
        view.closeEditor()
    }
}
